services/src/main/java/org/keycloak/protocol/oidc/endpoints/ParEndpoint.java

```diff
@@ -0,0 +1,76 @@
+package org.keycloak.protocol.oidc.endpoints;
+
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.annotations.cache.NoCache;
+import org.keycloak.authentication.ParResponse;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.models.CodeToTokenStoreProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.protocol.oidc.TokenManager;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.FormParam;
+import javax.ws.rs.POST;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+public class ParEndpoint {
+
+    private static final Logger LOG = Logger.getLogger(ParEndpoint.class);
+
+    public static final String OBJECT_ID = "objectId";
+    public static final String REDIRECT_ID = "redirectId";
+    public static final String OBJECT_TYPE = "objectType";
+    public static final String USERNAME = "username";
+    public static final String CLIENT_ID = "clientId";
+    public static final String REDIRECT_URI = "redirect_uri";
+
+    private final KeycloakSession session;
+    private final TokenManager tokenManager;
+    private final RealmModel realm;
+    private final EventBuilder event;
+
+    public ParEndpoint(KeycloakSession session, TokenManager tokenManager, RealmModel realm, EventBuilder event) {
+        this.session = session;
+        this.tokenManager = tokenManager;
+        this.realm = realm;
+        this.event = event;
+    }
+
+    @POST
+    @NoCache
+    @Consumes({MediaType.APPLICATION_FORM_URLENCODED})
+    public Response handlePar(@FormParam(OBJECT_ID) String objectId,
+                              @FormParam(REDIRECT_ID) String redirectId,
+                              @FormParam(OBJECT_TYPE) String objectType,
+                              @FormParam(USERNAME) String username,
+                              @FormParam(REDIRECT_URI) String redirectUri,
+                              @FormParam(CLIENT_ID) final String clientId) {
+
+        LOG.info("Received PAR for object ID: " + objectId);
+
+        CodeToTokenStoreProvider codeStore = session.getProvider(CodeToTokenStoreProvider.class,
```

Comment:
For the way of using `CodeToTokenStoreProvider`, it might be preferable to follow `OAuth2DeviceTokenStoreProvider` and `InfinispanOAuth2DeviceTokenStoreProvider` used for Device Grant.

Reply:
I have a question about this comment. The `OAuth2DeviceTokenStoreProvider ` uses much more functionality in comparison with the `CodeToTokenStoreProvider `. Basically, all we need is `put` and `get` methods, which are implemented in the `CodeToTokenStoreProvider `, we can store the map of parameters and then read this map only once (it is actually disposable). Do we really need the OAuth2DeviceTokenStoreProvider class?

Reply:
Yes, you are right. We need not to re-use `OAuth2DeviceTokenStoreProvider`.

IMO it might be better to implement PAR specific Store Provider which receive and return the POJO instance of PAR request and its key is `request_uri`.
```
    void put(String requestUri, int lifespanSeconds, ParRequestModel parRequest);
    ParRequestModel remove(String requestUri);
```
WDYT?

Reply:
Thank you for the advice! Yes, I have created a separate storage, factory and SPI for handling it. Those changes are already pushed to this branch, please, have a look.